codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -349,17 +350,68 @@ public String doubleShape(DoubleShape shape) {
     @Override
     public String unionShape(UnionShape shape) {
         writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
-        return """
-            func () types.%s {
-                var union types.%s
-                union = &types.MyUnionMemberIntegerValue{
-                    Value: %s.(int32),
-                }
-                return union
+        String functionSignature = "func() %s {".formatted(context.symbolProvider().toSymbol(shape));
+        String nilCheck = """
+                if %s == nil {
+                        return nil
+                }""".formatted(
+                        dataSource
+                );
+        if (GoPointableIndex.of(context.model()).isPointable(shape) == false) {
+            nilCheck = "";
+        }
+        String unionVariableInitialization = "var union %s".formatted(context.symbolProvider().toSymbol(shape));
```

Comment:
1. This whole thing can be simplified as:

```
"""
func() %s {
    %s
var union %s
"""
```
and so on.

2. finalize variables wherever required, you can even just do `final var xy = 1;`
3. Use `StringBuilder` when you are constructing strings.
4. Avoid `+` in strings if they are big.

Reply:
Done